src/multicall.provider.ts

```diff
@@ -45,11 +45,24 @@ type GroupTryAllResult<T extends any[][]> = {
   [K in keyof T]: K extends number ? TryAllResult<T[K]> : T[K];
 };
 
+// removes Provider.all() so it can be properly overriden even though the generics don't match
+// also turns properties visibility from private into protected so it's easier do use them
+declare class OmittedProvider {
+  protected _provider: EthersProvider;
+  protected _multicallAddress: string;
+  constructor(provider: EthersProvider, chainId?: number);
+  init(): Promise<void>;
+  getEthBalance(address: string): any;
+}
+
 /**
  * Provider class heavily based on ethers-multicall, but supporting specifying a blockTag for the multicall contract
  * call and `tryAggregate`.
  */
-class MulticallProvider extends Provider {
+class MulticallProvider extends (Provider as unknown as new (
+  provider: EthersProvider,
+  chainId?: number
+) => OmittedProvider) {
```

Comment:
Not sure about what is happening here, can you elaborate a bit more.

I would like to know why the right part of the extend is the best we can do. Using `as unknown` there doesn't seem very logical for me.

Reply:
Yes. So, the thing here is this was a pretty difficult thing to fix. The problem was there was an incompatibility between the generic method `all` (incompatibility in the template itself) between `MulticallProvider` and `Provider`, and considering messing with the base class is not an option, I had to force-cast `Provider` to a class declaration that didn't include that generic (it's only a change in TS realm, not JS).
Another option would be to use something like `@ts-ignore` or `@ts-expect-error` in `MulticallProvider.all()` declaration, but this would be worse since this would just ignore that declaration - effectively being as if that template didn't exist.

Reply:
There were other places I didn't like having to force-cast too. Maybe we should remove templating and just use the default `ethers.utils.Result` in relevant places? Let me know what you think.